File: php/src/ClientTypeAdapter.php
<?php

declare(strict_types=1);

namespace Apstndb\SpanValue;

final class ClientTypeAdapter
{
    /**
     * Adapt a client wrapper type to a wire `Type` shape (protojson-compatible keys).
     *
     * @return array<string, mixed>
     */
    public static function adapt(mixed $clientType): array
    {
        if ($clientType === null) {
            return ['code' => 'TYPE_CODE_UNSPECIFIED'];
        }

        $code = Proto::typeCode($clientType);
        $name = TypeCode::nameOf($code);
        /** @var array<string, mixed> $wire */
        $wire = ['code' => $name ?? $code];

        $ann = Proto::typeAnnotation($clientType);
        if ($ann !== TypeAnnotationCode::TYPE_ANNOTATION_CODE_UNSPECIFIED->value) {
            $annName = TypeAnnotationCode::nameOf($ann);
            if ($annName !== null) {
                $wire['typeAnnotation'] = $annName;
            }
        }

        $fqn = Proto::protoTypeFqn($clientType);
        if ($fqn !== '') {
            $wire['protoTypeFqn'] = $fqn;
        }

        $elem = Proto::arrayElementType($clientType);
        if ($elem !== null) {
            $wire['arrayElementType'] = self::adapt($elem);
        }

        if (Proto::structType($clientType) !== null) {
            $fields = [];
            foreach (Proto::structFields($clientType) as $field) {
                $fieldName = Proto::fieldName($field);
                /** @var array<string, mixed> $out */
                $out = [];
                if ($fieldName !== '') {
                    $out['name'] = $fieldName;
                }
                $out['type'] = self::adapt(Proto::fieldType($field));
                $fields[] = $out;
            }
            $wire['structType'] = ['fields' => $fields];
        }

        return $wire;
    }
}

File: ruby/lib/spanvalue/client_type_adapter.rb
# frozen_string_literal: true

module Spanvalue
  module ClientTypeAdapter
    module_function

    def adapt(client_type)
      return { code: 'TYPE_CODE_UNSPECIFIED' } if client_type.nil?

      code = Proto.type_code(client_type)
      name = Codes.type_code_name(code)
      wire = { code: name || code }

      ann = Proto.type_annotation(client_type)
      if !ann.nil? && ann != TypeAnnotationCode::TYPE_ANNOTATION_CODE_UNSPECIFIED
        wire[:typeAnnotation] = TypeAnnotationCode::NAMES[ann] || ann
      end

      fqn = Proto.proto_type_fqn(client_type)
      wire[:protoTypeFqn] = fqn unless fqn.empty?

      elem = Proto.array_element_type(client_type)
      wire[:arrayElementType] = adapt(elem) unless elem.nil?

      st = Proto.struct_type(client_type)
      unless st.nil?
        wire[:structType] = {
          fields: Proto.struct_fields(client_type).map do |field|
            out = { type: adapt(Proto.field_type(field)) }
            name = Proto.field_name(field)
            out[:name] = name unless name.empty?
            out
          end
        }
      end

      wire
    end
  end
end
